File: _includes/code/java-v6/src/test/java/SearchHybridTest.java
import io.weaviate.client6.v1.api.WeaviateClient;
import io.weaviate.client6.v1.api.collections.CollectionHandle;
import io.weaviate.client6.v1.api.collections.query.GroupBy;
import io.weaviate.client6.v1.api.collections.query.Hybrid.FusionType;
import io.weaviate.client6.v1.api.collections.query.Metadata;
import io.weaviate.client6.v1.api.collections.query.NearVector;
import io.weaviate.client6.v1.api.collections.query.Where;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.Map;

class SearchHybridTest {

  private static WeaviateClient client;

  @BeforeAll
  public static void beforeAll() throws IOException {
    // START INSTANTIATION-COMMON
    // Best practice: store your credentials in environment variables
    String weaviateUrl = System.getenv("WEAVIATE_URL");
    String weaviateApiKey = System.getenv("WEAVIATE_API_KEY");
    String openaiApiKey = System.getenv("OPENAI_APIKEY");

    client = WeaviateClient.connectToWeaviateCloud(
        weaviateUrl,
        weaviateApiKey,
        config -> config.setHeaders(Map.of("X-OpenAI-Api-Key", openaiApiKey)));
    // END INSTANTIATION-COMMON
  }

  @AfterAll
  public static void afterAll() throws Exception {
    client.close();
  }

  @Test
  void testNamedVectorHybrid() {
    // START NamedVectorHybridPython
    CollectionHandle<Map<String, Object>> reviews = client.collections.use("WineReviewNV");
    var response = reviews.query.hybrid(
        // TODO[g-despot] Why isn't targetVector available?
        // highlight-start
        "A French Riesling",
        q -> q
            // .targetVector("title_country")
            .limit(3)

    // highlight-end
    );

    for (var o : response.objects()) {
      System.out.println(o.properties());
    }
    // END NamedVectorHybridPython
  }

  @Test
  void testHybridBasic() {
    // START HybridBasicPython
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        // highlight-start
        "food",
        q -> q.limit(3)
    // highlight-end
    );

    for (var o : response.objects()) {
      System.out.println(o.properties());
    }
    // END HybridBasicPython
  }

  @Test
  void testHybridWithScore() {
    // START HybridWithScorePython
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        "food",
        q -> q
            .alpha(0.5f)
            // highlight-start
            .returnMetadata(Metadata.SCORE, Metadata.EXPLAIN_SCORE)
            // highlight-end
            .limit(3));

    for (var o : response.objects()) {
      System.out.println(o.properties());
      // highlight-start
      System.out.println(o.metadata().score() + " " + o.metadata().explainScore());
      // highlight-end
    }
    // END HybridWithScorePython
  }

  @Test
  void testLimit() {
    // START limit Python
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        "food",
        q -> q
            // highlight-start
            .limit(3)
            .offset(1)
    // highlight-end
    );

    for (var o : response.objects()) {
      System.out.println(o.properties());
    }
    // END limit Python
  }

  @Test
  void testAutocut() {
    // START autocut Python
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        "food",
        q -> q
            // highlight-start
            .fusionType(FusionType.RELATIVE_SCORE)
            .autocut(1)
    // highlight-end
    );

    for (var o : response.objects()) {
      System.out.println(o.properties());
    }
    // END autocut Python
  }

  @Test
  void testHybridWithAlpha() {
    // START HybridWithAlphaPython
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        "food",
        q -> q
            // highlight-start
            .alpha(0.25f)
            // highlight-end
            .limit(3));

    for (var o : response.objects()) {
      System.out.println(o.properties());
    }
    // END HybridWithAlphaPython
  }

  @Test
  void testHybridWithFusionType() {
    // START HybridWithFusionTypePython
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        "food",
        q -> q
            // highlight-start
            .fusionType(FusionType.RELATIVE_SCORE)
            // highlight-end
            .limit(3));

    for (var o : response.objects()) {
      System.out.println(o.properties());
    }
    // END HybridWithFusionTypePython
  }

  @Test
  void testHybridWithBM25OperatorOrWithMin() {
    // START HybridWithBM25OperatorOrWithMin
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        // highlight-start
        "Australian mammal cute"
    // TODO what about bm25Operator?
    // .bm25Operator(BM25Operator.or(2))
    // highlight-end
    // .limit(3)
    );

    for (var o : response.objects()) {
      System.out.println(o.properties());
    }
    // END HybridWithBM25OperatorOrWithMin
  }

  @Test
  void testHybridWithBM25OperatorAnd() {
    // START HybridWithBM25OperatorAnd
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        // highlight-start
        "Australian mammal cute"
    // TODO what about bm25Operator?
    // .bm25Operator(BM25Operator.and()) // Each result must include all tokens
    // (e.g. "australian", "mammal", "cute")
    // highlight-end
    // .limit(3)
    );

    for (var o : response.objects()) {
      System.out.println(o.properties());
    }
    // END HybridWithBM25OperatorAnd
  }

  @Test
  void testHybridWithProperties() {
    // START HybridWithPropertiesPython
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        "food",
        q -> q
            // highlight-start
            .queryProperties("question")
            // highlight-end
            .alpha(0.25f)
            .limit(3));

    for (var o : response.objects()) {
      System.out.println(o.properties());
    }
    // END HybridWithPropertiesPython
  }

  @Test
  void testHybridWithPropertyWeighting() {
    // START HybridWithPropertyWeightingPython
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        "food",
        q -> q
            // highlight-start
            .queryProperties("question^2", "answer")
            // highlight-end
            .alpha(0.25f)
            .limit(3));

    for (var o : response.objects()) {
      System.out.println(o.properties());
    }
    // END HybridWithPropertyWeightingPython
  }

  @Test
  void testHybridWithVector() {
    // START HybridWithVectorPython
    float[] queryVector = new float[1536]; // Some vector that is compatible with object vectors
    for (int i = 0; i < queryVector.length; i++) {
      queryVector[i] = -0.02f;
    }

    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        "food",
        q -> q
            // highlight-start
            .nearVector(NearVector.of(queryVector))
            // highlight-end
            .alpha(0.25f)
            .limit(3));

    for (var o : response.objects()) {
      System.out.println(o.properties());
    }
    // END HybridWithVectorPython
  }

  @Test
  void testHybridWithFilter() {
    // START HybridWithFilterPython
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        "food",
        q -> q
            // highlight-start
            .where(Where.property("round").eq("Double Jeopardy!"))
            // highlight-end
            .limit(3));

    for (var o : response.objects()) {
      System.out.println(o.properties());
    }
    // END HybridWithFilterPython
  }

  @Test
  void testVectorParameters() {
    // START VectorParametersPython
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        "California",
        q -> q
            // highlight-start
            // TODO[g-despot] Why is distance not available?
            // TODO[g-despot] Is there a simpler syntax?
            // .distance(0.4f) // Maximum threshold for the vector search component
            .nearVector(NearVector.of(jeopardy.query.nearText("large animal", c -> c
                .moveAway(0.5f, from -> from.concepts("mammal", "terrestrial")))
                .objects().get(0).vectors().getDefaultSingle()))
            // highlight-end
            .alpha(0.75f)
            .limit(5));
    // END VectorParametersPython
  }

  @Test
  void testVectorSimilarity() {
    // START VectorSimilarityPython
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        "California",
        q -> q
            // highlight-start
            // TODO[g-despot] Why is distance not available?
            // .distance(0.4f) // Maximum threshold for the vector search component
            // highlight-end
            .alpha(0.75f)
            .limit(5));
    // END VectorSimilarityPython
  }

  @Test
  void testHybridGroupBy() {
    // START HybridGroupByPy4
    // Query
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.query.hybrid(
        "California",
        q -> q.alpha(0.75f),
        GroupBy.property(
            "round", // group by this property
            2, // maximum number of groups
            3 // maximum objects per group
        ));

    response.groups().forEach((groupName, group) -> {
      System.out.println(group.name() + " " + group.objects());
    });
    // END HybridGroupByPy4
  }
}

File: _includes/code/csharp/SearchHybridTest.cs
using Xunit;
using Weaviate.Client;
using Weaviate.Client.Models;
using System;
using System.Threading.Tasks;
using System.Text.Json;
using System.Linq;
using System.Collections.Generic;

namespace WeaviateProject.Tests;

public class SearchHybridTest : IDisposable
{
    private static readonly WeaviateClient client;

    // Static constructor for one-time setup (like @BeforeAll)
    static SearchHybridTest()
    {
        // START INSTANTIATION-COMMON
        // Best practice: store your credentials in environment variables
        string weaviateUrl = Environment.GetEnvironmentVariable("WEAVIATE_URL");
        string weaviateApiKey = Environment.GetEnvironmentVariable("WEAVIATE_API_KEY");
        string openaiApiKey = Environment.GetEnvironmentVariable("OPENAI_APIKEY");

        client = Connect.Cloud(
                    weaviateUrl,
                    weaviateApiKey,
                    headers: new Dictionary<string, string>()
                    {
                { "X-OpenAI-Api-Key", openaiApiKey }
                    }
                ).GetAwaiter().GetResult();
        // END INSTANTIATION-COMMON
    }

    // Dispose is called once after all tests in the class are finished (like @AfterAll)
    public void Dispose()
    {
        // The C# client manages connections automatically and does not require an explicit 'close' method.
        GC.SuppressFinalize(this);
    }

    // TODO[g-despot] NEW: Grpc.Core.RpcException : Status(StatusCode="Unknown", Detail="extract target vectors: class WineReviewNV has multiple vectors, but no target vectors were provided")
    [Fact]
    public async Task NamedVectorHybrid()
    {
        // START NamedVectorHybrid
        var reviews = client.Collections.Use("WineReviewNV");
        // highlight-start
        var response = await reviews.Query.Hybrid(
            "A French Riesling",
            targetVector: ["title_country"],
            limit: 3
        );
        // highlight-end

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
        }
        // END NamedVectorHybrid

        Assert.Equal("WineReviewNV", response.Objects.First().Collection);
    }

    [Fact]
    public async Task TestHybridBasic()
    {
        // START HybridBasic
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            // highlight-start
            "food",
            limit: 3
        // highlight-end
        );

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
        }
        // END HybridBasic

        Assert.Equal("JeopardyQuestion", response.Objects.First().Collection);
    }

    [Fact]
    public async Task TestHybridWithScore()
    {
        // START HybridWithScore
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            "food",
            alpha: 0.5f,
            // highlight-start
            returnMetadata: MetadataOptions.Score | MetadataOptions.ExplainScore,
            // highlight-end
            limit: 3
        );

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
            // highlight-start
            Console.WriteLine($"Score: {o.Metadata.Score}, Explain Score: {o.Metadata.ExplainScore}");
            // highlight-end
        }
        // END HybridWithScore

        Assert.Equal("JeopardyQuestion", response.Objects.First().Collection);
        Assert.NotNull(response.Objects.First().Metadata.Score);
        Assert.NotNull(response.Objects.First().Metadata.ExplainScore);
    }

    [Fact]
    public async Task TestLimit()
    {
        // START limit 
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            "food",
            // highlight-start
            limit: 3,
            offset: 1
        // highlight-end
        );

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
        }
        // END limit 

        Assert.Equal("JeopardyQuestion", response.Objects.First().Collection);
        Assert.Equal(3, response.Objects.Count());
    }

    [Fact]
    public async Task TestAutocut()
    {
        // START autocut 
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            "food",
            // highlight-start
            fusionType: HybridFusion.RelativeScore,
            autoLimit: 1
        // highlight-end
        );

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
        }
        // END autocut 

        Assert.True(response.Objects.Any());
        Assert.Equal("JeopardyQuestion", response.Objects.First().Collection);
    }

    [Fact]
    public async Task TestHybridWithAlpha()
    {
        // START HybridWithAlpha
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            "food",
            // highlight-start
            alpha: 0.25f,
            // highlight-end
            limit: 3
        );

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
        }
        // END HybridWithAlpha

        Assert.Equal("JeopardyQuestion", response.Objects.First().Collection);
    }

    [Fact]
    public async Task TestHybridWithFusionType()
    {
        // START HybridWithFusionType
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            "food",
            // highlight-start
            fusionType: HybridFusion.RelativeScore,
            // highlight-end
            limit: 3
        );

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
        }
        // END HybridWithFusionType

        Assert.Equal("JeopardyQuestion", response.Objects.First().Collection);
    }

    [Fact]
    public async Task HybridWithBM25OperatorOr()
    {
        // START HybridWithBM25OperatorOrWithMin
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            // highlight-start
            "Australian mammal cute",
            bm25Operator: new BM25Operator.Or(MinimumMatch: 1),
            // highlight-end
            limit: 3
        );

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
        }
        // END HybridWithBM25OperatorOrWithMin

        Assert.Equal("JeopardyQuestion", response.Objects.First().Collection);
    }

    [Fact]
    public async Task HybridWithBM25OperatorAnd()
    {

        // START HybridWithBM25OperatorAnd
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            // highlight-start
            "Australian mammal cute",
            bm25Operator: new BM25Operator.And(),  // Each result must include all tokens
                                                   // highlight-end
            limit: 3
        );

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
        }
        // END HybridWithBM25OperatorAnd

        Assert.Equal("JeopardyQuestion", response.Objects.First().Collection);
    }

    [Fact]
    public async Task TestHybridWithProperties()
    {
        // START HybridWithProperties
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            "food",
            // highlight-start
            queryProperties: ["question"],
            // highlight-end
            alpha: 0.25f,
            limit: 3
        );

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
        }
        // END HybridWithProperties

        Assert.Equal("JeopardyQuestion", response.Objects.First().Collection);
    }

    [Fact]
    public async Task TestHybridWithPropertyWeighting()
    {
        // START HybridWithPropertyWeighting
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            "food",
            // highlight-start
            queryProperties: ["question^2", "answer"],
            // highlight-end
            alpha: 0.25f,
            limit: 3
        );

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
        }
        // END HybridWithPropertyWeighting

        Assert.Equal("JeopardyQuestion", response.Objects.First().Collection);
    }

    [Fact]
    public async Task TestHybridWithVector()
    {
        // START HybridWithVector
        var queryVector = Enumerable.Repeat(-0.02f, 1536).ToArray();

        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            "food",
            // highlight-start
            vectors: Vectors.Create(queryVector),
            // highlight-end
            alpha: 0.25f,
            limit: 3
        );

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
        }
        // END HybridWithVector

        Assert.Equal("JeopardyQuestion", response.Objects.First().Collection);
    }

    [Fact]
    public async Task TestHybridWithFilter()
    {
        // START HybridWithFilter
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            "food",
            // highlight-start
            filters: Filter.Property("round").Equal("Double Jeopardy!"),
            // highlight-end
            limit: 3
        );

        foreach (var o in response.Objects)
        {
            Console.WriteLine(JsonSerializer.Serialize(o.Properties));
        }
        // END HybridWithFilter

        Assert.Equal("JeopardyQuestion", response.Objects.First().Collection);
        Assert.Equal("Double Jeopardy!", (response.Objects.First().Properties as IDictionary<string, object>)["round"].ToString());
    }

    [Fact]
    public async Task TestVectorParameters()
    {
        // START VectorParameters
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        // This query is complex and depends on a previous nearText query to get a vector.
        // We simulate this by fetching a vector first.
        var nearTextResponse = await jeopardy.Query.NearText(
            "large animal",
            moveAway: new Move(force: 0.5f, concepts: ["mammal", "terrestrial"]),
            limit: 1,
            includeVectors: true
        );
        var nearTextVector = nearTextResponse.Objects.First().Vectors["default"];

        var response = await jeopardy.Query.Hybrid(
            "California",
            // highlight-start
            maxVectorDistance: 0.4f,
            vectors: nearTextVector,
            // highlight-end
            alpha: 0.75f,
            limit: 5
        );
        // END VectorParameters

        Assert.True(response.Objects.Any() && response.Objects.Count() <= 5);
    }

    [Fact]
    public async Task TestVectorSimilarity()
    {
        // START VectorSimilarity
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            "California",
            // highlight-start
            maxVectorDistance: 0.4f, // Maximum threshold for the vector search component
                                     // highlight-end
            alpha: 0.75f,
            limit: 5
        );
        // END VectorSimilarity

        Assert.True(response.Objects.Any() && response.Objects.Count() <= 5);
    }

    [Fact]
    public async Task TestHybridGroupBy()
    {
        // START HybridGroupByPy4
        // Query
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Query.Hybrid(
            "California",
            alpha: 0.75f,
            groupBy: new GroupByRequest
            {
                PropertyName = "round",       // group by this property
                NumberOfGroups = 2,           // maximum number of groups
                ObjectsPerGroup = 3,          // maximum objects per group
            }
        );

        foreach (var group in response.Groups.Values)
        {
            Console.WriteLine($"{group.Name} {JsonSerializer.Serialize(group.Objects)}");
        }
        // END HybridGroupByPy4

        Assert.True(response.Groups.Count > 0 && response.Groups.Count <= 2);
    }
}